### PR TITLE
[SW-673] max amount & fee calculator

### PR DESCRIPTION
**Pull request by ifaouibadi**

### CODE REVIEW

File: src/composables/maxAmount.ts [hunk 1]
```diff
@@ -45,111 +41,93 @@ export interface MaxAmountOptions extends IDefaultComposableOptions {
...
-        selectedAssetDecimals.value = val.selectedAsset.decimals!;
-      }
+    const maxValue = selectedAssetBalance.minus(fee.value);
+    return (maxValue.isPositive() ? maxValue : 0).toString();
+  });
 
-      const numericAmount = (val.amount && val.amount > 0) ? val.amount : 0;
-      const amount = new BigNumber(numericAmount).shiftedBy(MAGNITUDE);
+  async function updateCalculatedFee() {
+    if (!formModel.value?.selectedAsset) return;
```
Comment (pstachel) [inline src/composables/maxAmount.ts:69]:
```suggestion
    if (!formModel.value?.selectedAsset) {
       return;
     }
```

File: src/composables/maxAmount.ts [hunk 2]
```diff
@@ -45,111 +41,93 @@ export interface MaxAmountOptions extends IDefaultComposableOptions {
...
+    const sdk = await getSdk();
+    const numericAmount = (
+      formModel.value.amount
+      && formModel.value.amount > 0
```
Comment (pstachel) [inline src/composables/maxAmount.ts:73]:
```suggestion
      formModel.value?.amount > 0
```

Comment (pstachel) [inline src/composables/maxAmount.ts:73]:
```suggestion
      formModel.value?.amount > 0
```

File: src/composables/maxAmount.ts [hunk 3]
```diff
@@ -45,111 +41,93 @@ export interface MaxAmountOptions extends IDefaultComposableOptions {
...
-              : val.selectedAsset.contractId,
-          },
-        );
-        return;
-      }
+    try {
+      nonce.value = (await sdk.api
+        .getAccountByPubkey(account.value.address))?.nonce;
+    } catch (error: any) {
+      if (!error.message.includes('Account not found')) handleUnknownError(error);
```
Comment (pstachel) [inline src/composables/maxAmount.ts:81]:
```suggestion
      if (!error.message.includes('Account not found')) {
          handleUnknownError(error);
      }
```

File: src/composables/maxAmount.ts [hunk 4]
```diff
@@ -45,111 +41,93 @@ export interface MaxAmountOptions extends IDefaultComposableOptions {
...
+        ...sdk.Ae.defaults,
+        senderId: account.value.address,
+        recipientId: account.value.address,
+        amount,
+        ttl: 0,
+        nonce: nonce.value + 1,
+        payload: formModel.value.payload,
+      },
+    })).shiftedBy(-MAGNITUDE);
+    if (!minFee.isEqualTo(fee.value)) fee.value = minFee;
```
Comment (pstachel) [inline src/composables/maxAmount.ts:119]:
```suggestion
    if (!minFee.isEqualTo(fee.value)) {
        fee.value = minFee;
    }
```

**Review by pstachel — APPROVED** [review]

File: src/composables/maxAmount.ts [hunk 5]
```diff
@@ -45,111 +41,96 @@ export interface MaxAmountOptions extends IDefaultComposableOptions {
...
-          || tokenInstance.value.deployInfo.address !== val.selectedAsset.contractId
-        ) {
-          tokenInstance.value = await sdk.getContractInstance({
-            source: FUNGIBLE_TOKEN_CONTRACT,
-            contractAddress: val.selectedAsset.contractId,
-          });
-        }
-        selectedAssetDecimals.value = val.selectedAsset.decimals!;
-      }
+    const maxValue = selectedAssetBalance.minus(fee.value);
```
Comment (CedrikNikita) [inline src/composables/maxAmount.ts:64]: There are no need to minus fee if you are sending token.

**Review by CedrikNikita — CHANGES_REQUESTED** [review]

**Review by CedrikNikita — CHANGES_REQUESTED** [review]
- The reason for the constant updating of the token balance is that middleware is syncing slower than contract updates it's balance.
So sending two transactions in a row: first with amount and second one with MAX amount would fail in this case.
- The reason for fetching nonce is to calculate fee correctly (because at some point nonce will be high enough to impact the fee)

**Pull request closed without merging**